### ISSUE TITLE
如何对参数的比特进行求导

### ISSUE THREAD

**Issue by mumu029** [elliothe/Neural_Network_Weight_Attack#3]
你好，我阅读了你们的论文，非常感兴趣。但是有一个地方不太明白，想请教一下。我只学习过损失函数对模型参数的求导，但你们是如何实现损失函数对模型参数的每一比特求导的呢（我知道这里的参数被量化为了Nq个比特）？期待你的回答，谢谢。

Hello, I read your paper and am very interested. But there is a place I don't understand, I want to ask for advice. I have only learned how the loss function derives the model parameters, but how do you implement the loss function to derive each bit of the model parameters(I know the parameter here is quantized to Nq bits)? Looking forward to your answer, thank you.

![image](https://user-images.githubusercontent.com/51202355/200551273-d75f22f3-17cb-4434-837c-35d33957be1d.png)

**Comment by elliothe** [comment]
Please read our code for the answer. Calculate gradient w.r.t weight or weight bit is same for the autograd function in pytorch, as long as the gradient chain is not broken from the weight bits to weight.

**Comment by mumu029** [comment]
Thanks for all your responses on Github and in your email. If I ignore the concept of complement, am I right to understand this?
![image](https://user-images.githubusercontent.com/51202355/201461504-b1ae976f-a6d6-4217-948d-1ead32b478dc.png)

**Comment by elliothe** [comment]
In general, you understanding is correct. But the weight are encoded in two's complement thus the MSB is with -2^(i-1)
